File: math/partition_array_into_two_equal_product_subsets.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Boolean}
def check_equal_partitions(nums, target)
    # Calculate total product and length
    product = nums.reduce(:*)
    length = nums.length
    
    # Early return if product isn't target squared
    return false if product != target ** 2
    
    # Helper method for recursive checking
    def find_subset(nums, total, index, target, length)
        # Base cases
        return true if total == target
        return false if total > target || index == length
        
        # Try including or excluding current number
        find_subset(nums, total, index + 1, target, length) ||
        find_subset(nums, total * nums[index], index + 1, target, length)
    end
    
    # Start recursive search
    find_subset(nums, 1, 0, target, length)
end

File: math/partition_array_into_two_equal_product_subsets.py
class Solution:
    def checkEqualPartitions(self, nums: List[int], target: int) -> bool:
        product, length = 1, 0
        for num in nums:
            product *= num
            length += 1

        if product != target ** 2:
            return False

        def func(total, index):
            if total == target:
                return True
            if total > target or index == length:
                return False
            return func(total, index + 1) or func(total * nums[index], index + 1)

        return func(1, 0)